reference-library/BeforeSwapDelta.md

```diff
@@ -0,0 +1,218 @@
+|------------|--------------|--------------------------------------|
+| delta      | BeforeSwapDelta | The `BeforeSwapDelta` value       |
+
+Returns the extracted specified token delta as an `int128`.
+
+### getUnspecifiedDelta
+
+```solidity
+function getUnspecifiedDelta(BeforeSwapDelta delta) internal pure returns (int128 deltaUnspecified);
+```
+
+Extracts the unspecified token delta from a BeforeSwapDelta value.
+
+| Param Name | Type         | Description                          |
+|------------|--------------|--------------------------------------|
+| delta          | BeforeSwapDelta | The `BeforeSwapDelta`       |
+
+Returns the extracted unspecified token delta as an `int128`.
+
+## Usage in Hooks
+
+When a hook is called during a swap operation, it can perform custom logic and interact with the pool's balances. The `beforeSwap` hook returns a `BeforeSwapDelta` value to indicate any balance changes it introduces. This value can then be forwarded to the `afterSwap` hook, allowing for proper accounting and balance reconciliation.
+The `BeforeSwapDelta` can also be forwarded to the `afterAddLiquidity` and `afterRemoveLiquidity` hooks, enabling consistent balance tracking across various pool operations.
+
+## Usage in PoolManager.sol
+
+Here's how `BeforeSwapDelta` is used in the PoolManager contract:
+
+1. In the `swap` function of the PoolManager contract, the `beforeSwap` hook is called:
+
+```solidity
+function swap(PoolKey memory key, IPoolManager.SwapParams memory params, bytes calldata hookData)
+    external
+    override
+    onlyWhenUnlocked
+    noDelegateCall
+    returns (BalanceDelta swapDelta)
+{
+    // ... (other code)
+
+    BeforeSwapDelta beforeSwapDelta;
+    {
+        int256 amountToSwap;
+        uint24 lpFeeOverride;
+        (amountToSwap, beforeSwapDelta, lpFeeOverride) = key.hooks.beforeSwap(key, params, hookData);
+
+        // ... (swap execution)
+    }
+
+    // ... (other code)
+}
+```
+
+The `beforeSwap` hook returns a `BeforeSwapDelta` value along with other parameters.
+
+- This `beforeSwapDelta` can then be passed to the `afterSwap` hook:
+
+```solidity
+BalanceDelta hookDelta;
+(swapDelta, hookDelta) = key.hooks.afterSwap(key, params, swapDelta, hookData, beforeSwapDelta);
+```
+
+The `BeforeSwapDelta` serves several important purposes in this context:
+
+1. **Customization of Swap Behavior:** It allows hooks to modify the swap parameters or even completely override the default swap behavior.
+2. **Information Passing:** It provides a way for the `beforeSwap` hook to pass information to the `afterSwap` hook, enabling more complex and stateful hook logic.
+3. **Balance Adjustment:** The delta values can be used to adjust the final balance changes resulting from the swap, giving hooks fine-grained control over the swap's outcome.
+4. **Gas Optimization:** By packing two `int128` values into a single `int256`, it reduces the number of stack variables and can lead to gas savings.
+
```

Comment:
we should add hook fees here too. technically developers can charge fees on either side -- i think best practice is on unspecified afterSwap, but worth mentioning that beforeSwapDelta can be used for hook fees